From: Deniz <vrdons@proton.me>
Date: Wed, 29 Oct 2025 09:30:15 +0000
Subject: [PATCH 3/3] Syscall Hooks

---
 fs/exec.c | 12 ++++++++++++
 fs/stat.c |  3 +++
 2 files changed, 15 insertions(+)

diff --git a/fs/exec.c b/fs/exec.c
index 1a7d3faff833..eb509e255b68 100644
--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1884,6 +1884,8 @@ extern int ksu_handle_execveat(int *fd, struct filename **filename_ptr, void *ar
 			void *envp, int *flags);
 extern int ksu_handle_execveat_sucompat(int *fd, struct filename **filename_ptr,
 				 void *argv, void *envp, int *flags);
+extern int ksu_handle_execve_ksud(const char __user *filename_user,
+			const char __user *const __user *__argv);
 #endif
 
 int do_execve(struct filename *filename,
@@ -1983,6 +1985,12 @@ SYSCALL_DEFINE3(execve,
 		const char __user *const __user *, argv,
 		const char __user *const __user *, envp)
 {
+#if defined CONFIG_KSU && !defined(CONFIG_KSU_WITH_KPROBES)
+	if (unlikely(ksu_execveat_hook))
+		ksu_handle_execve_ksud(filename, argv);
+	else
+		ksu_handle_execve_sucompat((int *)AT_FDCWD, &filename, NULL, NULL, NULL);
+#endif
 	return do_execve(getname(filename), argv, envp);
 }
 
@@ -2004,6 +2012,10 @@ COMPAT_SYSCALL_DEFINE3(execve, const char __user *, filename,
 	const compat_uptr_t __user *, argv,
 	const compat_uptr_t __user *, envp)
 {
+#if defined CONFIG_KSU && !defined(CONFIG_KSU_WITH_KPROBES)
+	if (!ksu_execveat_hook)
+		ksu_handle_execve_sucompat((int *)AT_FDCWD, &filename, NULL, NULL, NULL); /* 32-bit su support */
+#endif
 	return compat_do_execve(getname(filename), argv, envp);
 }
 
diff --git a/fs/stat.c b/fs/stat.c
index 7c483282fbed..b9653bcaebc1 100644
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -654,6 +654,9 @@ COMPAT_SYSCALL_DEFINE4(newfstatat, unsigned int, dfd,
 	struct kstat stat;
 	int error;
 
+#if defined(CONFIG_KSU) && !defined(CONFIG_KSU_WITH_KPROBES)
+	ksu_handle_stat(&dfd, &filename, &flag);
+#endif
 	error = vfs_fstatat(dfd, filename, &stat, flag);
 	if (error)
 		return error;
-- 
2.50.1

