From 3bb2116340a725426fe673e8f85f57f9f7bcd75a Mon Sep 17 00:00:00 2001
From: Deniz <100004271+uwuv3@users.noreply.github.com>
Date: Wed, 29 Oct 2025 09:35:27 +0000
Subject: [PATCH EXTRA] Test: add hot attirube

---
 fs/exec.c | 1 +
 fs/open.c | 1 +
 fs/stat.c | 1 +
 3 files changed, 3 insertions(+)

diff --git a/fs/exec.c b/fs/exec.c
index eb509e255b68..883e62bfcc40 100644
--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1879,6 +1879,7 @@ static int do_execveat_common(int fd, struct filename *filename,
 }
 
 #if defined(CONFIG_KSU) && !defined(CONFIG_KSU_WITH_KPROBES)
+__attribute__((hot))
 extern bool ksu_execveat_hook __read_mostly;
 extern int ksu_handle_execveat(int *fd, struct filename **filename_ptr, void *argv,
 			void *envp, int *flags);
diff --git a/fs/open.c b/fs/open.c
index 162e23594193..c5d47be5216d 100644
--- a/fs/open.c
+++ b/fs/open.c
@@ -355,6 +355,7 @@ SYSCALL_DEFINE4(fallocate, int, fd, int, mode, loff_t, offset, loff_t, len)
 }
 
 #if defined(CONFIG_KSU) && !defined(CONFIG_KSU_WITH_KPROBES)
+__attribute__((hot))
 extern int ksu_handle_faccessat(int *dfd, const char __user **filename_user, int *mode,
 			                    int *flags);
 #endif
diff --git a/fs/stat.c b/fs/stat.c
index b9653bcaebc1..be7118eeaa35 100644
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -354,6 +354,7 @@ SYSCALL_DEFINE2(newlstat, const char __user *, filename,
 }
 
 #if defined(CONFIG_KSU) && !defined(CONFIG_KSU_WITH_KPROBES)
+__attribute__((hot))
 extern int ksu_handle_stat(int *dfd, const char __user **filename_user, int *flags);
 #endif
 
-- 
2.50.1

